src/block-cache.ts

```diff
@@ -0,0 +1,193 @@
+import {
+  createAsyncMiddleware,
+  JsonRpcMiddleware,
+} from 'json-rpc-engine';
+import {
+  cacheIdentifierForPayload,
+  blockTagForPayload,
+  cacheTypeForPayload,
+  canCache,
+  Payload,
+  Block,
+  BlockCache,
+  Cache,
+  JsonRPCRequestToCache,
+} from './cache-utils';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires,@typescript-eslint/no-require-imports
+const BlockTracker = require('eth-block-tracker');
+
+// `<nil>` comes from https://github.com/ethereum/go-ethereum/issues/16925
+const emptyValues = [undefined, null, '\u003cnil\u003e'];
+
+interface BlockCacheMiddlewareOptions{
+  blockTracker?: typeof BlockTracker;
+}
+
+export = createBlockCacheMiddleware;
+
+//
+// Cache Strategies
+//
+
+class BlockCacheStrategy {
+
+  private cache: Cache;
+
+  constructor() {
+    this.cache = {};
+  }
+
+  getBlockCacheForPayload(_payload: Payload, blockNumberHex: string): BlockCache {
+    const blockNumber: number = Number.parseInt(blockNumberHex, 16);
+    let blockCache: BlockCache = this.cache[blockNumber];
+    // create new cache if necesary
+    if (!blockCache) {
+      const newCache: BlockCache = {};
+      this.cache[blockNumber] = newCache;
+      blockCache = newCache;
+    }
+    return blockCache;
+  }
+
+  async get(payload: Payload, requestedBlockNumber: string): Promise<Block|undefined> {
+    // lookup block cache
+    const blockCache: BlockCache = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    // lookup payload in block cache
+    const identifier: string|null = cacheIdentifierForPayload(payload, true);
+    if (typeof identifier === 'string') {
+      return blockCache[identifier];
+    }
+    return undefined;
+  }
+
+  async set(payload: Payload, requestedBlockNumber: string, result: Block): Promise<void> {
+    // check if we can cached this result
+    const canCacheResult: boolean = this.canCacheResult(payload, result);
+    if (!canCacheResult) {
+      return;
+    }
+    const identifier: string|null = cacheIdentifierForPayload(payload, true);
+    if (!identifier) {
+      return;
+    }
+    // set the value in the cache
+    const blockCache: BlockCache = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    blockCache[identifier] = result;
```

Comment:
I originally asked that we add an early return here, which we should, but to be good, we should do that in a follow-up. We should do this for now, and I'll create an issue:
```suggestion
    // set the value in the cache
    const blockCache: BlockCache = this.getBlockCacheForPayload(payload, requestedBlockNumber);
    // TODO: Add early return if the identifier is falsy to avoid this cast
    blockCache[identifier as any] = result;
```

Reply:
Issue: #73